text/0043-storage-proof-size-hostfunction.md

```diff
@@ -0,0 +1,48 @@
+# RFC-0043: Introduce `storage_proof_size` Host Function for Improved Parachain Block Utilization
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 30 October 2023                                                                    |
+| **Description** | Host function to provide the storage proof size to runtimes.                                                                    |
+| **Authors**     | Sebastian Kunert                                                                                            |
+
+## Summary
+
+This RFC proposes a new host function for parachains, `storage_proof_size`.  It shall provide the size of the currently recorded storage proof to the runtime. Runtime authors can use the proof size to improve block utilization by retroactively reclaiming unused storage weight.
+
+## Motivation
+The number of extrinsics that are included in a parachain block is limited by two constraints: execution time and proof size. FRAME weights cover both concepts, and block-builders use them to decide how many extrinsics to include in a block. However, these weights are calculated ahead of time by benchmarking on a machine with reference hardware. The execution-time properties of the state-trie and its storage items are unknown at benchmarking time. Therefore, we make some assumptions about the state-trie:
+- **Trie Depth:** We assume a trie depth to account for intermediary nodes.
+- **Storage Item Size:** We make a pessimistic assumption based on the `MaxEncodedLen` trait.
+
+These pessimistic assumptions lead to an overestimation of storage weight, negatively impacting block utilization on parachains.
+
+In addition, the current model does not account for multiple accesses to the same storage items. While these repetitive accesses will not increase storage-proof size, the runtime-side weight monitoring will account for them multiple times. Since the proof size is completely opaque to the runtime, we can not implement retroactive storage weight correction.
+
+A solution must provide a way for the runtime to track the exact storage-proof size consumed on a per-extrinsic basis.
+
+## Stakeholders
+- **Parachain Teams:** They must include this host function in their runtime and node.
+- **Light-client Implementors:** They should include this host function in their runtime and node.
+
+## Explanation
+This RFC proposes a new host function that exposes the storage-proof size to the runtime. As a result, runtimes can implement storage weight reclaiming mechanisms that improve block utilization.
+
+This RFC proposes the following host function signature:
+```rust
+fn ext_storage_proof_size_version_1() -> Option<u64>;
```

Comment:
Returning an `Option` will cause the host to allocate 5 bytes of memory that the runtime will then free. This is terribly inefficient.
I would suggest returning `-1` rather than `None`.

Reply:
The goal was to get rid of the ambiguity between empty proof and proof recording disabled. 

> I would suggest returning -1 rather than None.
Assuming you mean `u32::MAX`, I am also fine with that :+1: .

Reply:
> Assuming you mean u64::MAX

Yes.
Substrate does some automagic interface adjustments, and under the hood this `u64` is actually a `i64`.